jerry-core/ecma/builtin-objects/ecma-builtin-json.c

```diff
@@ -628,48 +632,109 @@ ecma_builtin_json_internalize_property (ecma_object_t *reviver_p, /**< reviver f
   {
-    ecma_object_t *object_p = ecma_get_object_from_value (value);
+    /* 3.a */
+    ecma_value_t is_array = ecma_is_value_array (value);
 
-    ecma_collection_t *props_p = ecma_op_object_get_property_names (object_p, ECMA_LIST_ENUMERABLE);
+#if ENABLED (JERRY_ESNEXT)
+    if (ECMA_IS_VALUE_ERROR (is_array))
+    {
+      ecma_free_value (value);
+      return is_array;
+    }
+#endif /* ENABLED (JERRY_ESNEXT) */
 
-    ecma_value_t *buffer_p = props_p->buffer_p;
+    ecma_object_t *object_p = ecma_get_object_from_value (value);
 
-    /* 3.d.iii */
-    for (uint32_t i = 0; i < props_p->item_count; i++)
+    /* 3.c */
+    if (ecma_is_value_true (is_array))
     {
-      ecma_string_t *property_name_p = ecma_get_string_from_value (buffer_p[i]);
-
-      /* 3.d.iii.1 */
-      ecma_value_t result = ecma_builtin_json_internalize_property (reviver_p, object_p, property_name_p);
+      /* 3.c.ii */
+      uint32_t len;
+      ecma_value_t to_len = ecma_op_object_get_length (object_p, &len);
 
-      /* 3.d.iii.2 */
-      if (ECMA_IS_VALUE_ERROR (result))
+      /* 3.c.iii */
+      if (ECMA_IS_VALUE_ERROR (to_len))
       {
-        ecma_collection_free (props_p);
         ecma_deref_object (object_p);
-
-        return result;
+        return to_len;
       }
 
-      /* 3.d.iii.3 */
-      if (ecma_is_value_undefined (result))
+      /* 3.c.iv */
+      for (uint32_t i = 0; i < len; i++)
       {
-        ecma_value_t delete_val = ecma_op_general_object_delete (object_p,
-                                                                 property_name_p,
-                                                                 false);
-        JERRY_ASSERT (ecma_is_value_boolean (delete_val));
+        ecma_string_t *prop_index = ecma_new_ecma_string_from_uint32 (i);
+
+        ecma_value_t new_element = ecma_builtin_json_internalize_property (reviver_p, object_p, prop_index);
+
+        if (ECMA_IS_VALUE_ERROR (new_element))
+        {
+          ecma_deref_ecma_string (prop_index);
+          ecma_deref_object (object_p);
+          return new_element;
+        }
+
+        if (ecma_is_value_undefined (new_element))
+        {
+          ecma_value_t delete_val = ecma_op_object_delete_by_uint32_index (object_p,
+                                                                           i,
+                                                                           false);
+          JERRY_ASSERT (ecma_is_value_boolean (delete_val));
+        }
+        else
+        {
+          ecma_builtin_json_define_value_property (object_p,
+                                                   prop_index,
+                                                   new_element);
+          ecma_free_value (new_element);
+        }
+
+        ecma_deref_ecma_string (prop_index);
       }
-      /* 3.d.iii.4 */
-      else
+    }
+    /* 3.d */
+    else
+    {
+      ecma_collection_t *props_p = ecma_op_object_get_property_names (object_p, ECMA_LIST_ENUMERABLE);
```

Comment:
This operation might throw error, if `object_p` is a proxy object. Please add testcase for it as well.

Reply:
An assert is enough, because ```object_p``` can't be Proxy based on 24.5.1 step 6.
```
Assert: unfiltered is either a String, Number, Boolean, Null, or an Object that is defined by either an ArrayLiteral or an ObjectLiteral.
```